addons/web/static/src/js/views/graph/graph_renderer.js

```diff
@@ -1,1079 +1,1049 @@
+                        data.datasets.splice(originIndex, 0, {
+                            label: origin,
+                            data: fakeData,
+                            backgroundColor: colors.concat(['#d3d3d3']),
+                        });
+                        addNoDataToLegend = true;
+                    }
+                }
+                if (addNoDataToLegend) {
+                    data.labels.push(NO_DATA);
                 }
-            });
-            if (addNoDataToLegend) {
-                data.labels.push(NO_DATA);
             }
-        }
 
-        // prepare options
-        var options = this._prepareOptions(data.datasets.length);
-
-        // create chart
-        var ctx = document.getElementById(this.chartId);
-        this.chart = new Chart(ctx, {
-            type: 'pie',
-            data: data,
-            options: options,
-        });
-    },
-    /**
-     * Add the graph title (if any) above the canvas
-     *
-     * @private
-     */
-    _renderTitle: function () {
-        if (this.title) {
-            this.$el.prepend($('<label/>', {
-                text: this.title,
-            }));
-        }
-    },
-    /**
-     * Used to avoid too long legend items
-     *
-     * @private
-     * @param {string} label
-     * @returns {string} shortened version of the input label
-     */
-    _shortenLabel: function (label) {
-        // string returned could be 'wrong' if a groupby value contain a '/'!
-        var groups = label.split("/");
-        var shortLabel = groups.slice(0, 3).join("/");
-        if (shortLabel.length > 30) {
-            shortLabel = shortLabel.slice(0, 30) + '...';
-        } else if (groups.length > 3) {
-            shortLabel = shortLabel + '/...';
+            // prepare options
+            const options = this._prepareOptions(data.datasets.length);
+
+            // create chart
+            const ctx = this.canvasRef.el;
+            this.chart = new Chart(ctx, {
+                type: 'pie',
+                data: data,
+                options: options,
+            });
         }
-        return shortLabel;
-    },
-
-    //--------------------------------------------------------------------------
-    // Handlers
-    //--------------------------------------------------------------------------
-
-    /**
-     * @private
-     * @param {MouseEvent} ev
-     */
-    _onGraphClicked: function (ev) {
-        const activeElement = this.chart.getElementAtEvent(ev);
-        if (activeElement.length === 0) {
-            return;
+
+        /**
+         * Determine whether the data are good, and display an error message
+         * if this is not the case.
+         * @private
+         * @param {Object} props
+         */
+        _setNoContentHelper(props) {
+            this.noContentHelperData.show = false;
+
+            const dataPoints = this.filteredDataPoints;
+            if (!dataPoints.length && props.mode !== 'pie') {
```

Comment:
can't we first start with `if (props.mode == 'pie')` and simplify things?

Reply:
Of course! Do you feel my shame radiating to where you are? :/